### PR TITLE
Menu items list parent filter

### PR DESCRIPTION
**Pull request by tonypartridge**
Pull Request for Issue #16882  .

### Summary of Changes

Removes duplicate state filter
Added parent_id filtering

### Testing Instructions

Install patch, view list of menu items. Select search tools to show more search options and a new filter is on the right. called '- Select Parent Menu Item -'.

### CODE REVIEW

File: administrator/components/com_menus/models/forms/filter_items.xml [hunk 1]
```diff
@@ -74,6 +74,15 @@
 			>
 			<option value="">JOPTION_SELECT_MAX_LEVELS</option>
 		</field>
+		<field
+				name="parent_id"
```
Comment (Quy) [inline administrator/components/com_menus/models/forms/filter_items.xml]: Fix tab/spacing.

Comment (tonypartridge) [reply administrator/components/com_menus/models/forms/filter_items.xml]: It's late.. and drone didn't fail! :/ now fixed. Thanks!

File: administrator/language/en-GB/en-GB.com_menus.ini [hunk 2]
```diff
@@ -225,3 +225,6 @@ COM_MENUS_VIEW_NEW_ITEM_TITLE="Menus: New Item"
 COM_MENUS_VIEW_NEW_MENU_TITLE="Menus: Add"
 COM_MENUS_XML_DESCRIPTION="Component for creating menus."
 JLIB_RULES_SETTING_NOTES="Changes apply to this component only.<br /><em><strong>Inherited</strong></em> - a Global Configuration setting or higher level setting is applied.<br /><em><strong>Denied</strong></em> always wins - whatever is set at the Global or higher level and applies to all child elements.<br /><em><strong>Allowed</strong></em> will enable the action for this component unless it is overruled by a Global Configuration setting."
+COM_MENUS_FILTER_SELECT_PARENT_MENU_ITEM="- Select Parent Menu Item -"
+COM_MENUS_FILTER_PARENT_MENU_ITEM="Parent Menu Item"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_menus.ini]: COM_MENUS_FILTER_PARENT_MENU_ITEM_LABEL please

**Comment by coolcat-creations** [comment]
Cool thank you :) Will test it asap

**Comment by GeraintEdwards** [comment]
Test failed :(

Select 'Site Map' as parent_id and the list does indeed filter to show the correct subitems but the 'search tools disappear' - you need to add parent_id and a.parent_id to $config['filter_fields'] in the model constructor otherwise the 'search tools' disappear when you select a parent filter (with no other filters enabled).

Additionally it would be good for the parent item list to be filtered based on the main 'menu filter'.

**Comment by ghost** [comment]
@GeraintEdwards i altered your unsuccessfully Test at issue Tracker.

**Comment by tonypartridge** [comment]
Thanks @GeraintEdwards now resolved.

**Comment by GeraintEdwards** [comment]
I have tested this item :white_check_mark: successfully on ad24cdb5a4be800c33b375ba2dc49d17ad4ca857<br /><br />Menu items are filtered correctly and the filter remains visible in the 'search tools'<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17060">issues.joomla.org/tracker/joomla-cms/17060</a>.</sub>

**Comment by tonypartridge** [comment]
Now updated @coolcat-creations @GeraintEdwards to include a custom field type which dynamically filters the menu items by menu type if the menu items in the list are already filter by type.

**Comment by tonypartridge** [comment]
@rdeutz jenkis php70 docker isn't connecting?

**Comment by coolcat-creations** [comment]
I have tested this item :white_check_mark: successfully on 0b3ed6f31f26a51c43da8abda0e638a0a4218071<br /><br />Thank you, it works! Great improvement for large sites :)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17060">issues.joomla.org/tracker/joomla-cms/17060</a>.</sub>

**Comment by ghost** [comment]
@GeraintEdwards can you please retest?

**Comment by GeraintEdwards** [comment]
I have tested this item :white_check_mark: successfully on 0b3ed6f31f26a51c43da8abda0e638a0a4218071<br /><br />Filtering by menu type is a big functional improvement - thanks for adding it in.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17060">issues.joomla.org/tracker/joomla-cms/17060</a>.</sub>

**Comment by ghost** [comment]
RTC after two successful tests.

File: administrator/components/com_menus/models/forms/filter_items.xml (hunk 1 above)
Comment (C-Lodder) [reply administrator/components/com_menus/models/forms/filter_items.xml]: Drone is JS tests...don't be blaming drone :P

**Comment by rdeutz** [comment]
@tonypartridge you don't need to care about the new jenkins setup, we are still in testing mode

**Comment by tonypartridge** [comment]
That's fine, just didn't think it gets accepted unless all tests are passed.

Thanks!

**Comment by rdeutz** [comment]
travis and drone are required

File: administrator/components/com_menus/models/fields/menuitembytype.php [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+				$menuTitle = $menuType;
+			}
+
+			// Initialize the group.
+			$groups[$menuTitle] = array();
+
+			// Build the options array.
+			foreach ($items as $key => $link)
+			{
+
```
Comment (Quy) [inline administrator/components/com_menus/models/fields/menuitembytype.php]: Remove empty line.

File: administrator/language/en-GB/en-GB.com_menus.ini [hunk 4]
```diff
@@ -225,3 +225,6 @@ COM_MENUS_VIEW_NEW_ITEM_TITLE="Menus: New Item"
 COM_MENUS_VIEW_NEW_MENU_TITLE="Menus: Add"
 COM_MENUS_XML_DESCRIPTION="Component for creating menus."
 JLIB_RULES_SETTING_NOTES="Changes apply to this component only.<br /><em><strong>Inherited</strong></em> - a Global Configuration setting or higher level setting is applied.<br /><em><strong>Denied</strong></em> always wins - whatever is set at the Global or higher level and applies to all child elements.<br /><em><strong>Allowed</strong></em> will enable the action for this component unless it is overruled by a Global Configuration setting."
+COM_MENUS_FILTER_SELECT_PARENT_MENU_ITEM="- Select Parent Menu Item -"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.com_menus.ini]: Alpha order strings.

File: administrator/components/com_menus/models/fields/menuitembytype.php (hunk 3 above)
Comment (tonypartridge) [reply administrator/components/com_menus/models/fields/menuitembytype.php]: This is standard from the original menuitem.php file

**Comment by tonypartridge** [comment]
Made some slight changes as per @Quy 's request. Should not need retesting, nothing has changed apart from code style.

**Comment by Quy** [comment]
Actually, the lines should be moved after line 39.

File: administrator/components/com_menus/models/forms/filter_items.xml [hunk 5]
```diff
@@ -74,6 +74,15 @@
 			>
 			<option value="">JOPTION_SELECT_MAX_LEVELS</option>
 		</field>
+		<field
+			name="parent_id"
+			type="menuitembytype"
+			label="COM_MENUS_FILTER_PARENT_MENU_ITEM_LABEL"
+			description="COM_MENUS__FILTER_PARENT_MENU_DESC"
```
Comment (Quy) [inline administrator/components/com_menus/models/forms/filter_items.xml]: Fix typo. Use `COM_MENUS_FILTER_PARENT_MENU_ITEM_DESC`

File: administrator/language/en-GB/en-GB.com_menus.ini [hunk 6]
```diff
@@ -37,6 +37,9 @@ COM_MENUS_FIELD_VALUE_NEW_WITH_NAV="New Window With Navigation"
 COM_MENUS_FIELD_VALUE_NEW_WITHOUT_NAV="New Without Navigation"
 COM_MENUS_FIELD_VALUE_PARENT="Parent"
 COM_MENUS_FIELDSET_RULES="Permissions"
+COM_MENUS_FILTER_SELECT_PARENT_MENU_ITEM="- Select Parent Menu Item -"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.com_menus.ini]: These have to be in alpha order. 
```
COM_MENUS_FILTER_PARENT_MENU_ITEM_DESC
COM_MENUS_FILTER_PARENT_MENU_ITEM_LABEL
COM_MENUS_FILTER_SELECT_PARENT_MENU_ITEM
```

**Comment by tonypartridge** [comment]
@rdeutz has someone changed AppVeyor? This passed last week without issue, all I have done is moved language strings around in the .xml for @Quy ?

**Comment by rdeutz** [comment]
@tonypartridge it's nothing you can really count on, if they have failed I look into it and most of the time it is simply a wrong test result

**Comment by tonypartridge** [comment]
@rdeutz any chance we can get this merged for 3.7.4?

File: administrator/components/com_menus/models/fields/menuitembytype.php [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+ * @copyright   Copyright (C) 2005 - 2017 Open Source Matters, Inc. All rights reserved.
+ * @license     GNU General Public License version 2 or later; see LICENSE.txt
+ */
+
+defined('JPATH_PLATFORM') or die;
+
+JFormHelper::loadFieldClass('groupedlist');
+
+// Import the com_menus helper.
+require_once realpath(JPATH_ADMINISTRATOR . '/components/com_menus/helpers/menus.php');
```
Comment (mbabker) [inline administrator/components/com_menus/models/fields/menuitembytype.php]: Use `JLoader::register()` here.

File: administrator/components/com_menus/models/fields/menuitembytype.php [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+	 *                                      full field name would end up being "bar[foo]".
+	 *
+	 * @return  boolean  True on success.
+	 *
+	 * @see     JFormField::setup()
+	 * @since   3.2
+	 */
+	public function setup(SimpleXMLElement $element, $value, $group = null)
+	{
+		$app = JFactory::getApplication('administrator');
```
Comment (mbabker) [inline administrator/components/com_menus/models/fields/menuitembytype.php]: You can move this line down to inside the `if (!$menuType)` as you aren't always using the app.

File: administrator/components/com_menus/models/fields/menuitembytype.php [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+					{
+						$lang = ' (' . $link->language . ')';
+					}
+					else
+					{
+						$lang = '';
+					}
+
+					$groups[$menu->title][] = JHtml::_('select.option',
+						$link->value, $levelPrefix . $link->text . $lang,
```
Comment (mbabker) [inline administrator/components/com_menus/models/fields/menuitembytype.php]: If you're going to split the parameters across multiple lines, there should be one parameter per line (this line has two)

**Comment by tonypartridge** [comment]
I’ve just copied the core Joomla! Menuitem.php type……..

**Comment by tonypartridge** [comment]
Thanks @mbabker 👍

**Pull request merged**